### PR TITLE
MueLu: using kokkos for on node parallelism in Amalgamation process

### LINKED ISSUE #5838: MueLu: kokkos refactor of uncoupled aggregation and amalgamation process
## Enhancement
@trilinos/muelu 

This issue will track the progress made to integrate all changes from the branch [MueLu_uncoupled_aggregation_refactor](https://github.com/trilinos/Trilinos/tree/MueLu_uncoupled_aggregation_refactor). The aggregation refactor was done about a year ago and cannot be merged easily anymore so merging the previously mentioned branch will not be attempted.
Instead the changes in that branch will broken into smaller incremental changes with the hope that these incremental changes will be merged quickly and show a more logical progression.
As I tried to merge the old branch I also realized that some changes are potentially poorly designed and I am planning on modifying the work a bit to accommodate better design changes.
Here is a list of new features that are needed to complete the refactor work:

- [x] `MueLu::LWGraph` needs to new methods: `getRowPtrs()` and `getEntries()` to avoid awkward extraction of data from `graph_` see [here](https://github.com/trilinos/Trilinos/blob/21ec71443e69f66e2af1fad742bbdc8afb8411e7/packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase1Algorithm_kokkos_def.hpp#L280) and [here](https://github.com/trilinos/Trilinos/blob/21ec71443e69f66e2af1fad742bbdc8afb8411e7/packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase1Algorithm_kokkos_def.hpp#L329)
- [x] instead of modifying heavily the signature of `MueLu:: AggregationAlgorithmBase_kokkos::BuildAggregates(...)` I would like to update it to use a `Kokkos::View` instead of a `std::vector` for portability reasons and store pointers to additional data such as coloring information in `MueLu::Aggregates_kokkos`. This will avoid clashes with `StructuredAggregation` which does not require coloring information and already stores a pointer to a `MueLu::IndexManager_kokkos`
- [x] re-write the uncoupled aggregation phases using the work done in the MueLu_uncoupled_aggregation_refactor branch allowing for both non-deterministic and deterministic aggregation options
- [x] create `MueLu::AmalgamationInfo_kokkos` and `MueLu::AmalgamationFactory_kokkos` to set the infrastructure in MueLu for the related algorithms and data structured to be ported to kokkos.
- [ ] actually implement a new version of the amalgamation process that operates on `Kokkos::View` and `Aggregates_kokkos` and runs on device.
> Note: the current implementation does not allow the user to choose the coloring algorithm used by `UncoupledAggregationFactory_kokkos`. Currently the algorithm is set to `Serial` which means that the graph of the matrix is copied back to the CPU memory where the coloring is performed. While this approach provides a deterministic coloring we probably want to allow the user to run on GPU by specifying a different algorithm.

### PR DESCRIPTION
**Pull request by lucbv**
@trilinos/muelu 

## Motivation
Final step in the refactor of the aggregation process of MueLu.
After that profiling will be needed to assess performance and non-trivial serial parts of the code.

## Related Issues

* Closes #5838
* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of 

## Stakeholder Feedback
See issue #5838 

## Testing
Local testing has been performed as well as testing on `while`.

### CODE REVIEW

**Comment by lucbv** [comment]
@mhoemmen fair comment, to be honest this has been there for a long time, I am suspecting that guards should have been implemented to switch between the two definitions.
I do hope though that no-one is relying on this node definition as I have no idea how consistent it is with the rest of the code?
In any case I have decided to keep the public definition as it is most likely the one being used at the moment.

**Comment by lucbv** [comment]
@mhoemmen actually it is fine and somewhat necessary to have both definitions.
The first one just defines a short hand tpye: `node_type` and the second one creates an alias: "Node = node_type" that is being used internally for MueLu's short names header included just after the aliasing.

**Comment by mhoemmen** [comment]
@lucbv my bad -- thanks for checking!

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_AmalgamationFactory_kokkos_decl.hpp [hunk 1]
```diff
@@ -56,7 +56,7 @@
 #include "MueLu_Exceptions.hpp"
 
 #ifdef HAVE_MUELU_KOKKOS_REFACTOR
-#include "MueLu_AmalgamationInfo_kokkos_fwd.hpp"
+#include "MueLu_AmalgamationInfo_kokkos.hpp"
```
Comment (jhux2) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_AmalgamationFactory_kokkos_decl.hpp]: Doesn't the file `MueLu_AmalgamationInfo_kokkos.hpp` include this file, `MueLu_AmalgamationFactory_kokkos_decl.hpp`?

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_AmalgamationInfo_kokkos_decl.hpp [hunk 2]
```diff
@@ -86,8 +86,11 @@ namespace MueLu {
 
   public:
 
-    AmalgamationInfo_kokkos(RCP<Array<LO> > rowTranslation,
-                            RCP<Array<LO> > colTranslation,
+    using execution_space = typename LWGraph_kokkos::execution_space;
```
Comment (jhux2) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_AmalgamationInfo_kokkos_decl.hpp]: Does this need to be public?

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_AmalgamationFactory_kokkos_decl.hpp (hunk 1 above)
Comment (lucbv) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_AmalgamationFactory_kokkos_decl.hpp]: No, `MueLu_AmalgamationFactory_kokkos_{decl,def}.hpp` may include `MueLu_AmalgamationInfo_kokkos.hpp` but not the otherway around. Actually I think that `MueLu_AmalgamationInfo_kokkos_{decl,def}.hpp` have no reference to the `AmalgamationFactory_kokkos`.

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_AmalgamationInfo_kokkos_decl.hpp (hunk 2 above)
Comment (lucbv) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_AmalgamationInfo_kokkos_decl.hpp]: Maybe not but it is a type definition not a variable so I feel there is no harm in making it public?

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_AmalgamationFactory_kokkos_decl.hpp (hunk 1 above)
Comment (jhux2) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_AmalgamationFactory_kokkos_decl.hpp]: My mistake.  I misread `MueLu_AmalgamationInfo_kokkos` as `MueLu_AmalgamationFactory_kokkos`.

**Comment by jhux2** [comment]
@lucbv What would you like to do with this PR?

**Comment by lucbv** [comment]
Yeah, I'm not sure what it's state is to be honest...
I would actually lean toward closing it and reworking the issue if this is still something we need.

**Pull request closed without merging**